File: cp-library-rs/src/euler_tour.rs
use crate::consts::INF;

/// EulerTour
#[derive(Debug)]
pub struct EulerTour {
    pub G: Vec<Vec<usize>>,
    pub in_: Vec<usize>,
    pub out: Vec<usize>,
    pub depth: Vec<usize>,
}

impl EulerTour {
    /// 木を初期化する
    pub fn new(N: usize) -> Self {
        Self {
            G: vec![vec![]; N],
            in_: vec![INF; N],
            out: vec![INF; N],
            depth: vec![INF; N],
        }
    }

    /// 辺 `(u,v)` を追加する
    pub fn add_edge(&mut self, u: usize, v: usize) {
        self.G[u].push(v);
        self.G[v].push(u);
    }

    /// 順序付けを行う
    pub fn build(&mut self, root: usize) {
        self.dfs(INF, root, &mut 0, &mut 0);
    }

    /// 行きがけ順，帰りがけ順で順序付け
    fn dfs(&mut self, p: usize, u: usize, id: &mut usize, depth: &mut usize) {
        self.in_[u] = *id;
        self.depth[u] = *depth;

        *depth += 1;

        for i in 0..self.G[u].len() {
            let v = self.G[u][i];
            if v == p {
                continue;
            }
            *id += 1;
            self.dfs(u, v, id, depth);
        }

        *depth -= 1;
        *id += 1;

        self.out[u] = *id;
    }
}

File: cp-library-py/lib/graph/euler_tour.py
"""オイラーツアー"""


class EulerTour:
    def __init__(self, n: int):
        """頂点数 n で初期化する"""

        self.n = n
        self.G: list[list[int]] = [[] for _ in range(n)]
        self.in_: list[int] = []
        self.out: list[int] = []
        self.depth: list[int] = []

    def add_edge(self, u: int, v: int):
        """辺 {u,v} を追加する"""

        self.G[u].append(v)
        self.G[v].append(u)

    def build(self, root: int):
        """root を根として順序付けを行う"""

        in_ = [-1] * self.n
        out = [-1] * self.n
        depth = [-1] * self.n

        d, i = 0, 0
        stack = [root]

        while stack:
            u = stack.pop()
            if depth[u] == -1:
                stack.append(u)
                # 行きがけ処理
                in_[u] = i
                depth[u] = d
                i += 1
                d += 1
                for v in self.G[u]:
                    if depth[v] != -1:
                        continue
                    stack.append(v)
            else:
                # 帰りがけ処理
                d -= 1
                out[u] = i
                i += 1

        self.in_ = in_
        self.out = out
        self.depth = depth

    def get_id(self, v: int) -> tuple[int, int]:
        """頂点 v の (行きがけ順, 帰りがけ順) の番号を返す"""
        return (self.in_[v], self.out[v])
